### PR TITLE
fix: Fix resizing problems by replacing resizable with splitter controls

### LINKED ISSUE #3710: Edit mode does not resize properly when shrinking window after growing it
## Describe the bug

If you enlarge the composer window the edit area will grow to fill the space, but when the composer window is resized smaller the edit window stays the same size. If word wrap is enabled it remains stuck at the larger size and you cannot see / edit text that ends up off screen.

In addition Restart Bot / Test In emulator buttons stay off screen and it is impossible to scroll to them.  You have to leave edit mode for the UI to realign properly to make all controls / text visible again.

## Version

Bot Framework Composer
                
                Version:  1.0.2
                Electron: 8.2.4
                Chrome: 80.0.3987.165
                NodeJS: 12.13.0
                V8: 8.0.426.27-electron.0

## Browser

- [X] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [X] macOS
- [X] Windows
- [ ] Ubuntu

## To Reproduce

1. Go to Bot Responses
2. Select Edit mode
3. Enlarge window
4. Shrink window
5. Notice controls are clipped and text edit area is off screen / not resized

## Expected behavior

Expect UI elements to resize when composer window is shrunk to remain visible

## Screenshots

composerResizeBug

### LINKED ISSUE #4220: Replace resizer with react-splitter (embed code) in Composer

### PR DESCRIPTION
**Pull request by GeoffCoxMSFT**
## Description

This replaces using resizable with a splitter control.  Resizable is attached to a child control which causes resizing problems because it is not aware of the full area being split and does not property react to window sizes.  It also requires controls participating in a split to be aware of resizeable rather than being contained in a layout container control.

Verified both on web and within electron application.

## Task Item

Closes #4220 
Fixes #3710

## Screenshots

image

image

image

### CODE REVIEW

File: Composer/packages/client/src/components/NavTree.tsx [hunk 1]
```diff
@@ -3,23 +3,22 @@
 
 /** @jsx jsx */
 import { jsx, css } from '@emotion/core';
-import { Resizable, ResizeCallback } from 're-resizable';
+// import { Resizable, ResizeCallback } from 're-resizable';
```
Comment (hatpick) [inline Composer/packages/client/src/components/NavTree.tsx:6]: remove commented imports

File: Composer/packages/client/src/components/NavTree.tsx [hunk 2]
```diff
@@ -75,43 +74,44 @@ interface INavTreeProps {
 
 const NavTree: React.FC<INavTreeProps> = (props) => {
   const { navLinks, regionName } = props;
-  const { updateUserSettings } = useRecoilValue(dispatcherState);
-  const { dialogNavWidth: currentWidth } = useRecoilValue(userSettingsState);
+  // const { updateUserSettings } = useRecoilValue(dispatcherState);
```
Comment (hatpick) [inline Composer/packages/client/src/components/NavTree.tsx:77]: remove comments?

File: Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx [hunk 3]
```diff
@@ -16,13 +16,13 @@ import { FocusZone, FocusZoneDirection } from 'office-ui-fabric-react/lib/FocusZ
 import cloneDeep from 'lodash/cloneDeep';
 import formatMessage from 'format-message';
 import { DialogInfo, ITrigger } from '@bfc/shared';
-import { Resizable, ResizeCallback } from 're-resizable';
+// import { Resizable, ResizeCallback } from 're-resizable';
```
Comment (hatpick) [inline Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx:19]: Remove

File: Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx [hunk 4]
```diff
@@ -185,41 +183,43 @@ export const ProjectTree: React.FC<IProjectTreeProps> = (props) => {
     }
   };
 
-  const handleResize: ResizeCallback = (_e, _dir, _ref, d) => {
-    updateUserSettings({ dialogNavWidth: currentWidth + d.width });
-  };
+  // const handleResize: ResizeCallback = (_e, _dir, _ref, d) => {
```
Comment (hatpick) [inline Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx:186]: remove comments

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+const MeasureDiv = styled.div`
+  width: 100%;
+  height: 100%;
+  box-sizing: border-box;
+  outline: none;
+  overflow: hidden;
+`;
+
+const Root = styled.div(({ leftColWidth, splitterWidth }: { leftColWidth: string; splitterWidth: number }) => ({
```
Comment (hatpick) [inline Composer/packages/client/src/components/Split/LeftRightSplit.tsx:18]: why one is string the other one is number that will get converted to px string value later on L27?
They both can be either number or string right?

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    if (value.endsWith('px') || value.endsWith('%') || value.endsWith('fr')) {
+      return value;
+    }
+  }
+
+  return `${value}px`;
+};
+
+// constrains the extend of a pane given split constraints
+const constrainPaneExtent = (
```
Comment (hatpick) [inline Composer/packages/client/src/components/Split/LeftRightSplit.tsx:91]: Please add comments to the list of args

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+  // ensure within total bounds after adjustments
+  return Math.max(0, Math.min(newPrimary, total - splitter));
+};
+
+type Props = {
+  initialLeftGridWidth?: string | number;
+  minLeftPixels?: number;
+  minRightPixels?: number;
+  splitterWidth?: number;
+  renderSplitter?: () => JSX.Element;
```
Comment (hatpick) [inline Composer/packages/client/src/components/Split/LeftRightSplit.tsx:125]: Prefer React.ReactNode to JSX.Element

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+};
+
+export const LeftRightSplit = (props: React.PropsWithChildren<Props>): JSX.Element => {
```
Comment (hatpick) [inline Composer/packages/client/src/components/Split/LeftRightSplit.tsx:128]: nit: no need to define return type
React.ReactNode is better

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+  if (leftWidth < 0) {
+    renderLeftWidth = toGridWidth(defaultLeftWidth);
+  }
+
+  const children = React.Children.toArray(props.children);
+  const leftChild = children.length > 0 ? children[0] : <div />;
+  const rightChild = children.length > 1 ? children[1] : <div />;
+
+  const renderSplitVisual =
+    renderSplitter ??
```
Comment (hatpick) [inline Composer/packages/client/src/components/Split/LeftRightSplit.tsx:203]: When custom rendering the splitter, isn't the width of it important in your calculations? (L#27)

File: Composer/packages/client/src/components/Split/TopBottomSplit.tsx [hunk 10]
```diff
@@ -0,0 +1,232 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (hatpick) [inline Composer/packages/client/src/components/Split/TopBottomSplit.tsx:1]: What's your plan to merge the two splitter codes? (to get rid of the repetitive code in LR and TB splitters)

**Review by hatpick — CHANGES_REQUESTED** [review]

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx (hunk 5 above)
Comment (GeoffCoxMSFT) [reply Composer/packages/client/src/components/Split/LeftRightSplit.tsx:18]: The pane widths can be percent or pixels, but the splitter width can only be a pixel width.  I'll think about what I might want to do to allow pixel strings for the splitter width prop - future change.

File: Composer/packages/client/src/components/Split/LeftRightSplit.tsx (hunk 9 above)
Comment (GeoffCoxMSFT) [reply Composer/packages/client/src/components/Split/LeftRightSplit.tsx:203]: No - it should render within whatever area it is given by the splitterWidth just like panes do. The caller provided the splitterWidth param so they should have it for the custom render as well.

File: Composer/packages/client/src/components/Split/TopBottomSplit.tsx (hunk 10 above)
Comment (GeoffCoxMSFT) [reply Composer/packages/client/src/components/Split/TopBottomSplit.tsx:1]: I would do this in the OSS library and then port it here - merging the logic (which I did as a dry run) had a lot of impact on readability.

**Review by hatpick — DISMISSED** [review]

File: Composer/packages/client/src/pages/design/PropertyEditor.tsx [hunk 11]
```diff
@@ -28,11 +28,11 @@ const PropertyEditor: React.FC = () => {
   const { shellApi, ...shellData } = useShellApi();
   const { currentDialog, data: formData = {}, focusPath, focusedSteps, schemas } = shellData;
 
-  const currentWidth = shellData?.userSettings?.propertyEditorWidth || 400;
+  // const currentWidth = shellData?.userSettings?.propertyEditorWidth || 400;
```
Comment (cwhitten) [inline Composer/packages/client/src/pages/design/PropertyEditor.tsx:31]: what's the plan with this commented out stuff?

Comment (GeoffCoxMSFT) [reply Composer/packages/client/src/pages/design/PropertyEditor.tsx:31]: Dang - missed removing that one too :-S Will clean up.

**Review by cwhitten — APPROVED** [review]

**Pull request merged**